server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2352,58 +2343,121 @@ public void acquireReplicaOperationPermit(final long opPrimaryTerm, final long g
+                            if (currentGlobalCheckpoint < maxSeqNo) {
+                                resetEngineToGlobalCheckpoint();
+                            } else {
+                                getEngine().rollTranslogGeneration();
+                            }
                         });
                     }
                 }
             }
         }
-
         assert opPrimaryTerm <= pendingPrimaryTerm
-                : "operation primary term [" + opPrimaryTerm + "] should be at most [" + pendingPrimaryTerm + "]";
-        indexShardOperationPermits.acquire(
-                new ActionListener<Releasable>() {
-                    @Override
-                    public void onResponse(final Releasable releasable) {
-                        if (opPrimaryTerm < operationPrimaryTerm) {
-                            releasable.close();
-                            final String message = String.format(
-                                    Locale.ROOT,
-                                    "%s operation primary term [%d] is too old (current [%d])",
-                                    shardId,
-                                    opPrimaryTerm,
-                                    operationPrimaryTerm);
-                            onPermitAcquired.onFailure(new IllegalStateException(message));
-                        } else {
-                            assert assertReplicationTarget();
-                            try {
-                                updateGlobalCheckpointOnReplica(globalCheckpoint, "operation");
-                                advanceMaxSeqNoOfUpdatesOrDeletes(maxSeqNoOfUpdatesOrDeletes);
-                            } catch (Exception e) {
-                                releasable.close();
-                                onPermitAcquired.onFailure(e);
-                                return;
-                            }
-                            onPermitAcquired.onResponse(releasable);
-                        }
-                    }
+            : "operation primary term [" + opPrimaryTerm + "] should be at most [" + pendingPrimaryTerm + "]";
+    }
 
-                    @Override
-                    public void onFailure(final Exception e) {
-                        onPermitAcquired.onFailure(e);
+    /**
+     * Creates a new action listener which verifies that the operation primary term is not too old. If the given primary
+     * term is lower than the current one, the {@link ActionListener#onFailure(Exception)} method of the provided listener is invoked with
+     * an {@link IllegalStateException}. Otherwise the global checkpoint and the max_seq_no_of_updates marker of the replica are updated
+     * before the invocation of the {@link ActionListener#onResponse(Object)}} method of the provided listener.
+     */
+    private ActionListener<Releasable> createListener(final ActionListener<Releasable> listener,
```

Comment:
I'm not so comfortable with separating this code from the one in `updatePrimaryTermIfNeeded` - they are tightly connected. Instead of sharing code this way, how about creating a callback that will run:
```
indexShardOperationPermits.acquire(listener, executorOnDelay, true, debugInfo);
```
or 
```
indexShardOperationPermits.asyncBlockOperations(listener, timeout.duration(), timeout.timeUnit());
```

Reply:
I share your concerns and I updated the code, let me know what you think.